content/en/blog/_posts/2025-01-14-devices-failure-handling/index.md

```diff
@@ -0,0 +1,409 @@
+* The pod watcher relies on external controllers to reschedule a Pod.
+
+There are more variations of DIY solutions for handling device failures or
+upcoming maintenance. Overall, Kubernetes has enough extension points to
+implement these solutions. However, some extension points require higher
+privilege than users may be comfortable with or are too disruptive. The roadmap
+section goes into more details on specific improvements in handling the device
+failures.
+
+### Failure modes: container code failed
+
+When the container code fails or something bad happens with it, like out of
+memory conditions, Kubernetes knows how to handle those cases. There is either
+the restart of a container, or a crash of a Pod if it has `restartPolicy: Never`
+and scheduling it on another node. Kubernetes has limited expressiveness on what
+is a failure (for example, non-zero exit code or liveness probe failure) and how
+to react on such a failure (mostly either Always restart or immediately fail the
+Pod).
+
+This level of expressiveness is often not enough for the complicated AI/ML
+workloads. AI/ML pods are better rescheduled locally or even in-place as that
+would save on image pulling time and device allocation. AI/ML pods are often
+interconnected and need to be restarted together. This adds another level of
+complexity and optimizing it often brings major savings in running AI/ML
+workloads.
+
+There are various DIY solutions to handle Pod failures orchestration. The most
+typical one is to wrap a main executable in a container by some orchestrator.
+And this orchestrator will be able to restart the main executable whenever the
+job needs to be restarted because some other pod has failed.
+
+Solutions like this are very fragile and elaborate. They are often worth the
+money saved comparing to a regular JobSet delete/recreate cycle when used in
+large training jobs. Making these solutions less fragile and more streamlined
+by developing new hooks and extension points in Kubernetes will make it
+easy to apply to smaller jobs, benefiting everybody.
+
+### Failure modes: device degradation
+
+Not all device failures are terminal for the overall workload or batch job.
+As the hardware stack gets more and more
+complex, misconfiguration on one of the hardware stack layers, or driver
+failures, may result in devices that are functional, but lagging on performance.
+One device that is lagging behind can slow down the whole training job.
+
+We see reports of such cases more and more often. Kubernetes has no way to
+express this type of failures today and since it is the newest type of failure
+mode, there is not much of a best practice offered by hardware vendors for
+detection and third party tooling for remediation of these situations.
+
+Typically, these failures are detected based on observed workload
+characteristics. For example, the expected speed of AI/ML training steps on
+particular hardware.
```

Comment:
and what would be the remediation in this case? I don't see what would be the ideal state here; should such nodes with degraded devices get restarted automatically? should we just give the taint on those nodes? should we just restart the training pods or device plugins? I know it could depend on degradations, but listing some of them could help readers imagining those.

Reply:
the remediation covered in a roadmap.

```suggestion
particular hardware. Remediation for those issues is highly depend on a workload needs.
```